Update and expand Hello Config (#410)
nextflow-io/training: docs/hello_nextflow/06_hello_config.md

TODO (refine)-> Learn to modify the execution of a pipeline _without altering a single line of code_. This highlights the power of Nextflow's configuration; enabling you to control how the pipeline runs without changing what it runs. Use this flexibility to adapt your pipeline to run in any environment.

## 0. Warmup: Moving to a formal project structure

So far we've been working with a very loose structure, with just one workflow code file and a tiny configuration file that we've mostly ignored, because we were very focused on learning how to implement the workflow itself.
However, we're now moving into the phase of this training series that is more focused on code development and maintenance practices.

As part of that, we're going to adopt a formal project structure. We're going to work inside a dedicated project directory called `projectC` (C for configuration), and we've renamed the workflow file `main.nf` to match the recommended Nextflow convention.

```console title="Directory contents"
projectC/
├── data -> ../data
├── main.nf
└── nextflow.config
```

> The `main.nf` file is a copy of the workflow produced by completing Part 4 of this training course.

Reviewer: Future proof this a bit? 
```suggestion
The `main.nf` file is similar to the workflow produced by completing Part 4 of this training course.
```

You know how to configure which software package each process should use, and how to switch between technologies.

### What's next?

Learn how to use profiles to make selecting an option easier.

---

## 2. Use profiles to select preset configurations

Profiles are a great way to adapt your workflow configuration by selecting preset options at runtime, to avoid having to edit a file every time you want to run something differently.

### 2.1. Create profiles for switching between Docker and Conda

Setting up these profiles mainly involves restructuring how we specify the `docker` and `conda` directives.

_Before:_

```groovy title="nextflow.config" linenums="1"
docker.enabled = false
conda.enabled = true

docker.fixOwnership = true
```

_After:_

```groovy title="nextflow.config" linenums="1"
docker.fixOwnership = true

profiles {
    docker_on {
        docker.enabled = true
    }
    conda_on {
        conda.enabled = true
    }
}
```

This makes it possible to activate one or the other by specifying a profile in our Nextflow run command line.

### 2.2. Run the workflow with a profile

Let's try running the workflow with Conda.

```bash
nextflow run main.nf -profile conda_on
```

> It works! And from our standpoint, it looks like it works exactly the same, even though on the backend the mechanics are a bit different.

Author: Oh, I meant that to say that the mechanics of using a conda environment are quite different compared to using a docker container but I realize how it reads; this is the wrong place to do it. Will edit.

### 3.2. Launch the workflow to generate the job submission script

Let's try running this; even though we now it won't execute (since we don't have Slurm set up in the Gitpod environment) we'll be able to see what the submission script looks like.

```bash
nextflow run main.nf -profile conda_on
```

As expected, this fails with a fairly unambiguous error:

```console title="Output"
nextflow
 N E X T F L O W   ~  version 24.02.0-edge

 ┃ Launching `main.nf` [grave_gauss] DSL2 - revision: 66cd7c255a

[-        ] SAMTOOLS_INDEX       [  0%] 0 of 3
[eb/2962ce] SAMTOOLS_INDEX (3)   [ 33%] 1 of 3, failed: 1
[-        ] GATK_HAPLOTYPECALLER -
[-        ] GATK_JOINTGENOTYPING -
ERROR ~ Error executing process > 'SAMTOOLS_INDEX (3)'

Caused by:
  java.io.IOException: Cannot run program "sbatch" (in directory "/workspace/gitpod/hello-nextflow/projectC/work/eb/2962ce167b3025a41ece6ce6d7efc2"): error=2, No such file or directory

Command executed:

  sbatch .command.run
```

However, it did produce what we are looking for: the `.command.run` file that Nextflow tried to submit to Slurm via the `sbatch` command.

> Let's take a look inside. **TODO: UPDATE NEXTFLOW VERSION SO WE CAN HAVE THIS SWEET OUTPUT**

Author: I plan to deal with this once the updates to Modules are done too. 

So will merge into dev with this still in, but I'm making a note that we have to address this before merging to master.

    Whenever there is a file named `nextflow.config` in the current directory, Nextflow will automatically load configuration from it. The one we have been using contains the following lines:

    ```console title="nextflow.config" linenums="1"
    docker.enabled = true

>     docker.fixOwnership = true

Author: We do want the `docker.fixOwnership = true` line up top so it's out of the way

    We'll cover both of those approaches in an upcoming part of this training course.

That being said, there may be some constraints on what you can (or must) allocate depending on what computing executor and compute infrastructure you're using. For example, your cluster may require you to stay within certain limits that don't apply when you're running elsewhere.

### 4.5. Add resource limits to an HPC profile

You can use the `resourceLimits` directive to set the relevant limitations. The syntax looks like this when it's by itself in a process block:

```groovy title="Syntax"
process {
    resourceLimits = [
        memory: 750.GB,
        cpus: 200,
        time: 30.d
    ]
}
```

Let's add this to the `univ_hpc` profile we set up earlier.

_Before:_

```groovy title="nextflow.config"
    univ_hpc {
        process.executor = 'slurm'
        conda.enabled = true
    }
```

_After:_

```groovy title="nextflow.config"
    univ_hpc {
        process.executor = 'slurm'
        conda.enabled = true
        process.resourceLimits = [
            memory: 750.GB,
            cpus: 200,
            time: 30.d
        ]
    }
```

> We can't test this since we don't have a live connection to Slurm in the Gitpod environment, but you can run it and look up the `sbatch` command in the `.command.run` script file if you'd like to see how these directives are translated into job parameters for the executor.

Author: So where can people look to see how these instructions are provided to Slurm?

> We can't test this since we don't have a live connection to Slurm in the Gitpod environment, but you can run it and look up the `sbatch` command in the `.command.run` script file if you'd like to see how these directives are translated into job parameters for the executor.

Author: Ohh I get it now. Because Nextflow isn't *telling* Slurm what the limits are, it's *enforcing* them before submitting jobs through `sbatch`, is that right? 

You can probably tell I've only run on cloud myself, not HPC 😅

Anyway, will edit. Thanks.

> We can't test this since we don't have a live connection to Slurm in the Gitpod environment, but you can run it and look up the `sbatch` command in the `.command.run` script file if you'd like to see how these directives are translated into job parameters for the executor.

Author: That makes sense, thank you!

> We can't test this since we don't have a live connection to Slurm in the Gitpod environment, but you can run it and look up the `sbatch` command in the `.command.run` script file if you'd like to see how these directives are translated into job parameters for the executor.

Author: Rephrased like this, lmk if that's still wrong.

> We can't test this since we don't have a live connection to Slurm in the Gitpod environment.
> However, you can try running the workflow with resource allocations that exceed these limits, then look up the `sbatch` command in the `.command.run` script file to see what requests actually get sent to the executor.
